packages/NodeTypeResolver/src/NodeTypeResolver.php

```diff
@@ -15,27 +15,25 @@
 
     public function addPerNodeTypeResolver(PerNodeTypeResolverInterface $perNodeTypeResolver): void
     {
-        $this->perNodeTypeResolvers[$perNodeTypeResolver->getNodeClass()] = $perNodeTypeResolver;
+        foreach ($perNodeTypeResolver->getNodeTypes() as $nodeType) {
+            $this->perNodeTypeResolvers[$nodeType] = $perNodeTypeResolver;
+        }
     }
 
     /**
      * @return string[]
      */
     public function resolve(Node $node): array
     {
-        foreach ($this->perNodeTypeResolvers as $class => $perNodeTypeResolver) {
-            if (! $node instanceof $class) {
-                continue;
-            }
-
-            // resolve just once
-            if ($node->getAttribute(Attribute::TYPES)) {
-                return $node->getAttribute(Attribute::TYPES);
-            }
+        if (! isset($this->perNodeTypeResolvers[$node->getType()])) {
+            return [];
+        }
```

Comment:
I finally found use of https://github.com/nikic/PHP-Parser/commit/1c11626f0a2fda566d9db9f388feb0682093955a

This should be much faster than foreach

Reply:
Let’s see how it goes 🙏🏻